agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/profiler/PerformanceMonitoringService.java

```diff
@@ -0,0 +1,164 @@
+
+package com.microsoft.applicationinsights.agent.internal.profiler;
+
+import com.azure.monitor.opentelemetry.exporter.implementation.utils.ThreadPoolUtils;
+import com.microsoft.applicationinsights.agent.internal.configuration.Configuration;
+import com.microsoft.applicationinsights.agent.internal.profiler.service.ServiceProfilerClient;
+import com.microsoft.applicationinsights.agent.internal.profiler.triggers.AlertingSubsystemInit;
+import com.microsoft.applicationinsights.agent.internal.profiler.upload.UploadService;
+import com.microsoft.applicationinsights.agent.internal.profiler.util.ServiceLoaderUtil;
+import com.microsoft.applicationinsights.agent.internal.telemetry.TelemetryClient;
+import com.microsoft.applicationinsights.agent.internal.telemetry.TelemetryObservers;
+import com.microsoft.applicationinsights.alerting.AlertingSubsystem;
+import com.microsoft.applicationinsights.alerting.config.AlertingConfiguration;
+import com.microsoft.applicationinsights.diagnostics.DiagnosticEngine;
+import com.microsoft.applicationinsights.diagnostics.DiagnosticEngineFactory;
+import java.io.File;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This holds the various components of the monitoring system:
+ *
+ * <ul>
+ *   <li>Profiler
+ *   <li>Alerting subsystem
+ *   <li>Profile Upload service
+ *   <li>Diagnostic engine
+ * </ul>
+ */
+public class PerformanceMonitoringService {
+  private static final Logger logger = LoggerFactory.getLogger(PerformanceMonitoringService.class);
+  private final String processId;
+  private final String machineName;
+  private final String roleName;
+  private final TelemetryClient telemetryClient;
+  private final Configuration configuration;
+  private final File tempDir;
+
+  private boolean currentlyEnabled = false;
+
+  //////////////////////////////////////////////////////////
+  // These are brought up and down on demand if the profiler is fully enabled
+  @Nullable private AlertingSubsystem alerting;
+  @Nullable private UploadService uploadService;
+  @Nullable private ScheduledExecutorService alertServiceExecutorService;
+  @Nullable private Profiler profiler;
+  @Nullable private DiagnosticEngine diagnosticEngine;
+  @Nullable private ScheduledExecutorService diagnosticEngineExecutorService;
+  //////////////////////////////////////////////////////////
+
+  public PerformanceMonitoringService(
+      String processId,
+      String machineName,
+      String roleName,
+      TelemetryClient telemetryClient,
+      Configuration configuration,
+      File tempDir) {
+    this.processId = processId;
+    this.machineName = machineName;
+    this.roleName = roleName;
+    this.telemetryClient = telemetryClient;
+    this.configuration = configuration;
+    this.tempDir = tempDir;
+  }
+
+  synchronized void enableProfiler(
```

Comment:
Large synchornized block.  Do we need to sync all of this logic?  Would explicit Object locks be easier to understand (probably not)

Reply:
we could tighten it quite a bit, main thing to protect is `currentlyEnabled` to prevent double initialization

Reply:
let's add a comment to revisit synchronization once we add "disablement"